libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -1263,7 +1263,39 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
 				// as require with custom errors is not supported in legacy codegen.
 				auto const* magicType = dynamic_cast<MagicType const*>(arguments[1]->annotation().type);
 				if (magicType && magicType->kind() == MagicType::Kind::Error)
-					solUnimplemented("Require with a custom error is only available using the via-ir pipeline.");
+				{
+					// Make sure that error constructor arguments are evaluated regardless of the require condition
+					auto const& errorConstructorCall = dynamic_cast<FunctionCall const&>(*arguments[1]);
+					errorConstructorCall.expression().accept(*this);
+					std::vector<Type const*> argumentTypes{};
+					unsigned stackDepth{0};
+					for (ASTPointer<Expression const> const& arg: errorConstructorCall.sortedArguments())
+					{
+						arg->accept(*this);
+						stackDepth += arg->annotation().type->sizeOnStack();
+						argumentTypes.push_back(arg->annotation().type);
+					}
+					// stack: <condition> <arg0> <arg1> ... <argN>
+					// Move condition to the top of the stack
+					utils().moveIntoStack(arguments.at(0)->annotation().type->sizeOnStack(), stackDepth);
+					// stack: <arg0> <arg1> ... <argN> <condition>
+					m_context << Instruction::ISZERO << Instruction::ISZERO;
+					auto success = m_context.appendConditionalJump();
+
+					auto const* errorDefinition = dynamic_cast<ErrorDefinition const*>(ASTNode::referencedDeclaration(errorConstructorCall.expression()));
+					solAssert(errorDefinition && errorDefinition->functionType(true));
+
+					utils().revertWithError(
+						errorDefinition->functionType(true)->externalSignature(),
+						errorDefinition->functionType(true)->parameterTypes(),
+						argumentTypes
+					);
+					m_context.adjustStackOffset(static_cast<int>(stackDepth));
```

Comment:
Maybe a comment similar to the one in the revert-with-string case as an explanation for this is helpful?
```
// Here, the arguments are consumed, but in the other branch, they are still there.
```
or so

Reply:
Oh, and what about this one? It was marked as resolved but without any answer and the comment is still there.

Reply:
It doesn't consume the arguments, and if it did, adjusting of offset would break things. `abiEncode` is called within `revertWithError`, which is what consumes the error constructor call arguments - the same is not present in the success branch, and thus the same arguments are left on the stack - we even have a test for this behaviour: `test/libsolidity/semanticTests/errors/require_error_stack_check.sol`.

Reply:
How is `popStackSlots(sizeOfErrorArguments)` not consuming the arguments? It inserts a series of `POP`s in the success branch and each `POP` consumes one item from the stack, doesn't it?

And I do see that tests pass so I'm not questioning whether it works. Just whether the explanation why it works is correct. I'd prefer to understand what's happening here.

Reply:
The comment applies to the place where the comment is not to code that occurs later. Of course the arguments are still there in the other branch - that's why you then pop them there.